plugin/pkg/auth/authenticator/password/saslauthd/saslauthd.go

```diff
@@ -0,0 +1,144 @@
+/*
+Copyright 2015 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package saslauthd
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/auth/authenticator"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/auth/user"
+	"net"
+	"os"
+	"strings"
+)
+
+type saslAuthdAuthenticator struct {
+	serviceName  string
+	socketPath   string
+	defaultRealm string
+}
+
+// NewSaslAuthd returns a password authenticator that checks passwords using
+// saslauthd by connecting directly to its socket.  The request format matches
+// the one expected by saslauthd versions 2.0.5 and later (through at least
+// 2.1.26): login ID, password, service name, and realm, each preceded by a
+// 16-bit network-order length, with no padding.
+func NewSaslAuthd(defaultRealm, socketPath, serviceName string) (authenticator.Password, error) {
+	if serviceName == "" {
+		serviceName = os.Args[0]
+	}
+	index := strings.LastIndex(serviceName, "/")
+	if index >= 0 {
+		serviceName = serviceName[index+1:]
+	}
+	if serviceName == "" {
+		return nil, errors.New("error determining SASL service name")
+	}
+	return saslAuthdAuthenticator{serviceName: serviceName, socketPath: socketPath, defaultRealm: defaultRealm}, nil
+}
+
+// AuthenticatePassword implements authenticator.Password by handing it off to
+// saslauthd, using the specified username as the login ID.  If successful, a
+// user.DefaultInfo is returned.  That structure will contain the username, an
+// empty UID, and no group information.
+func (s saslAuthdAuthenticator) AuthenticatePassword(username, password string) (user.Info, bool, error) {
+	var replen uint16
+	realm := s.defaultRealm
+	index := strings.LastIndex(username, "@")
+	if index >= 0 {
+		realm = username[index+1:]
+		username = username[0:index]
+	}
+	if username == "" {
+		return nil, false, errors.New("No user name specified for basic authentication")
+	}
+	if password == "" {
+		return nil, false, errors.New("No password specified for basic authentication")
```

Comment:
would an empty password ever be allowed? if so, this code shouldn't decide it's an error, but should let the sasl service fail it